design/one-pager-oam-workflow.md

```diff
@@ -0,0 +1,228 @@
+# OAM POC Workflow
+* Owner: Dan Mangum (@hasheddan)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+This document describes a basic workflow for utilizing OAM types as implemented
+in [1276](https://github.com/crossplane/crossplane/pull/1276). The controllers
+described are for an initial POC and will be moved out of core Crossplane in
+later iterations.
+
+## 1. Crossplane Installation
+
+The following OAM CRDs will be installed with Crossplane:
+
+- `ApplicationConfiguration`
+- `Component`
+- `TraitDefinition`
+- `ScopeDefinition`
+- `WorkloadDefinition`
+- `ContainerizedWorkload`
+
+Crossplane will also create a `WorkloadDefinition` *instance* that looks as
+follows:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: WorkloadDefinition
+metadata:
+  name: containerizedworkloads.core.oam.dev
+spec:
+  definitionRef:
+    name: containerizedworkloads.core.oam.dev
+```
+
+This makes the `ContainerizedWorkload` core workload type available to be used
+by OAM `Components`. Other workload types can be created and made available by
+creating a CRD for the workload and a `WorkloadDefinition` that references it.
+
+The OAM-related controllers will also be started on installation:
+
+- **Application Configuration controller**: watches `ApplicationConfiguration`
+  resources and creates the corresponding workloads based on the inline
+  parameters and referenced `Components`
+- **Containerized Workload controller**: watches `ContainerizedWorkload`
+  resources and packages them into `KubernetesApplications` for scheduling to a
+  target cluster 
+
+## 2. Component Creation
+
+Users create a `Component` to define a unit of deployment that can be included
+in an `ApplicationConfiguration`. A `Component` must specify a
+`WorkloadDefinition`.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: Component
+metadata:
+  name: myapp
+spec:
+  workload:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    metadata:
+      name: my-workload
+    spec:
+      osType: linux
+      containers:
+      - name: tbs11-app
+        image: hasheddan/tbs11:latest
+        resources:
+          cpu:
+            required: 1.0
+          memory:
+            required: 100MB
+        env:
+          - name: DB_HOST
+            valueFrom:
+              secretKeyRef:
+                name: mysqlconn
+                key: endpoint
```

Comment:
Does `ContainerizedWorkload` support `valueFrom`?

Reply:
It does not currently in #1276 and the implementation there differs from the current [OAM Spec](https://github.com/oam-dev/spec/blob/master/3.component_model.md#env) as it has a `fromParam` field (although it is marked as not required). If we do end up adopting the full k8s definition of a `Container` then these fields would be present.